src/rdkafka_broker.c

```diff
@@ -3631,13 +3633,62 @@ static void rd_kafka_toppar_fetch_backoff (rd_kafka_broker_t *rkb,
                 return;
 
         rktp->rktp_ts_fetch_backoff = rd_clock() + (backoff_ms * 1000);
+
         rd_rkb_dbg(rkb, FETCH, "BACKOFF",
-                   "%s [%"PRId32"]: Fetch backoff for %dms: %s",
+                   "%s [%"PRId32"]: Fetch backoff for %dms%s %s",
                    rktp->rktp_rkt->rkt_topic->str, rktp->rktp_partition,
-                   backoff_ms, rd_kafka_err2str(err));
+                   backoff_ms,
+                   err ? ":" : "", 
+                   err ? rd_kafka_err2str(err) : "");
 }
 
 
+/**
+ * @brief Handle preferred replica in fetch response.
+ */
+void
+rd_kafka_fetch_preferred_replica_handle (rd_kafka_toppar_t *rktp,
+                                         rd_kafka_buf_t *rkbuf,
+                                         rd_kafka_broker_t *rkb,
+                                         int32_t preferred_id) {
+        if (rd_interval_immediate(&rktp->rktp_new_lease_intvl,
+                                  60*1000000/*1 minute*/, 0) < 0) {
+                rd_rkb_log(rkb, LOG_WARNING, "FETCH",
+                        "%.*s [%"PRId32"]: Not migrating to preferred "
+                        "replica %"PRId32" because previous lease was "
+                        "granted < 1 minute ago",
+                        RD_KAFKAP_STR_PR(rktp->rktp_rkt->rkt_topic),
+                        rktp->rktp_partition, preferred_id);
+
+                rd_kafka_toppar_fetch_backoff(rkb,
+                        rktp, RD_KAFKA_RESP_ERR_NO_ERROR);
+        } else {
+                rd_kafka_broker_t *preferred_rkb;
+                rd_kafka_t *rk = rktp->rktp_rkt->rkt_rk;
+
+                rd_kafka_rdlock(rk);
+                preferred_rkb = rd_kafka_broker_find_by_nodeid(rk,
+                                                               preferred_id);
+                rd_kafka_rdunlock(rk);
+
+                if (preferred_rkb) {
+                        rd_kafka_topic_wrlock(rktp->rktp_rkt);
+                        rd_interval_reset_to_now(&rktp->rktp_lease_intvl, 0);
```

Comment:
the `rkt` lock is there because it's required by `rd_kafka_toppar_broker_update` -> `rd_kafka_toppar_broker_delegate`. It would be possible to refactor the function structure in order to minimize how long the lock is held c.f. the current code, but it would be more verbose and I think the difference would be inconsequential.

you're right, no lock is required for the lease interval, and i've moved this up a line.